daemon/inspect_fs_unix_fstab.ml

```diff
@@ -523,24 +523,54 @@ and resolve_fstab_device spec md_map os_type =
     default
   )
 
-and resolve_dev_mapper spec default =
-  (* LVM2 does some strange munging on /dev/mapper paths for VGs and
-   * LVs which contain '-' character:
-   *
-   * ><fs> lvcreate LV--test VG--test 32
-   * ><fs> debug ls /dev/mapper
-   * VG----test-LV----test
-   *
-   * This makes it impossible to reverse those paths directly, so
-   * we have implemented lvm_canonical_lv_name in the daemon.
-   *)
-  try
-    match Lvm_utils.lv_canonical spec with
-    | None -> default
-    | Some device -> Mountable.of_device device
-  with
-  (* Ignore devices that don't exist. (RHBZ#811872) *)
-  | Unix.Unix_error (Unix.ENOENT, _, _) -> default
+and resolve_crypttab spec aug =
+  let augpath =
+    sprintf "/files/etc/crypttab/*[target='%s']/device"
+            (Filename.basename spec) in
+  match aug_get_noerrors aug augpath with
+  | None -> None
+  | Some device ->
+     if verbose() then eprintf "mapped to crypttab device=%s\n%!" device;
+     (* device string is one of:
+      *  + UUID=... without any shell quoting
+      *  + An absolute path
+      *)
+      if String.starts_with "UUID=" device then (
+        (* Map the UUID to the /dev/mapper naming that `-i` will
+         * use by default with cryptsetup-open. But if that doesn't
+         * exist, just return the path from fstab.
+         *)
+        let uuid = String.sub device 5 (String.length device - 5) in
+        let guestfs_path = sprintf "/dev/mapper/luks-%s" uuid in
+        if Sys.file_exists guestfs_path then
+          Some(Mountable.of_device guestfs_path)
+        else
+          Some(Mountable.of_device spec)
+      ) else (
+        Some(Mountable.of_device spec)
+      )
+
+and resolve_dev_mapper spec default aug =
+    match resolve_crypttab spec aug with
```

Comment:
I wonder if this code would be more straightforward if `resolve_crypttab` was inlined.  But I wouldn't press the point.

Reply:
Ok I moved it inline but it lead to the whole LVM block being indented which makes the diff quite larger. I think I'm missing something. If hypothetically resolve_dev_mapper grew more /dev/mapper logic would we chain it inside successively nested `match` blocks?

Reply:
There's no real rule, whatever would be best in that scenario.